File: tatala-java/src/com/qileyuan/tatala/socket/client/LongClientSession.java
package com.qileyuan.tatala.socket.client;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.lang.reflect.Method;
import java.net.BindException;
import java.net.InetSocketAddress;
import java.net.StandardSocketOptions;
import java.nio.ByteBuffer;
import java.nio.channels.AsynchronousSocketChannel;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.LinkedTransferQueue;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.zip.DataFormatException;

import org.apache.log4j.Logger;

import com.qileyuan.tatala.proxy.DefaultProxy;
import com.qileyuan.tatala.socket.exception.SocketExecuteException;
import com.qileyuan.tatala.socket.exception.TatalaRollbackException;
import com.qileyuan.tatala.socket.to.StandardTransferObject;
import com.qileyuan.tatala.socket.to.TransferObject;
import com.qileyuan.tatala.socket.util.TransferUtil;

public class LongClientSession{
	static Logger log = Logger.getLogger(LongClientSession.class);
	static final int BUFFER_SIZE = 1024;
	static final int QUEUE_SIZE = 10;
	
	private String ip;
	private int port;
	private int timeout;
	
	private AsynchronousSocketChannel socketChannel;
	private AioClientReceiveHandler aioClientReceiveHandler = new AioClientReceiveHandler();
	private BlockingQueue<byte[]> receiveQueue = new LinkedTransferQueue<byte[]>();
	private DefaultProxy serverCallProxy;
	
	private ByteBuffer byteBuffer = ByteBuffer.allocate(BUFFER_SIZE);
	private ByteArrayOutputStream bos = new ByteArrayOutputStream();
	private boolean handleFlag = false;
	private boolean firsttime = true;
	private int expectReceiveLength = 0;
	private int receiveLength = 0;
	private boolean closed = false;
	
	public LongClientSession(String ip, int port, int timeout){
		this.ip = ip;
		this.port = port;
		this.timeout = timeout;
	}
	
	public Object start(TransferObject to) {
		Object resultObject = null;
		String calleeClass = to.getCalleeClass();
		String calleeMethod = to.getCalleeMethod();
		
		//set default server call proxy
		if(to.getServerCallProxy() != null){
			serverCallProxy = to.getServerCallProxy();
		}
		
		try {
			if(socketChannel == null || !socketChannel.isOpen() || closed){
				connect();
			}
			send(to);
			resultObject = receive(to);
			
			if(resultObject instanceof TatalaRollbackException){
				throw (TatalaRollbackException)resultObject;
			}
			
		} catch (TatalaRollbackException tre) {
			log.error("Tatala Return Exception: Callee Class and Method: [" + calleeClass + "." + calleeMethod + "]");
			throw tre;
		} catch (BindException be) {
			log.error("Connection error: " + be.getMessage());
			log.error("Callee Class and Method: [" + calleeClass + "." + calleeMethod + "]");
			close();
		} catch (TimeoutException te) {
			log.error("Socekt timed out, return null. [" + timeout + "ms]");
			log.error("Callee Class and Method: [" + calleeClass + "." + calleeMethod + "]");
			close();
		} catch (Exception e) {
			log.error("Callee Class and Method: [" + calleeClass + "." + calleeMethod + "]");
			e.printStackTrace();
			close();
		}
		
		return resultObject;
	}
	
	private void connect() throws BindException{
		String errorMessage = "";
		if (socketChannel == null || !socketChannel.isOpen() || closed) {
			try {
				socketChannel = AsynchronousSocketChannel.open();
				socketChannel.setOption(StandardSocketOptions.SO_REUSEADDR, true);
				socketChannel.setOption(StandardSocketOptions.SO_KEEPALIVE, true);
				socketChannel.connect(new InetSocketAddress(ip, port)).get(timeout, TimeUnit.MILLISECONDS);
				closed = false;
				log.debug("Session start to " + socketChannel.getRemoteAddress());
				
				//when connect to the server, keep receiving data either server response or server call
				receive();
			} catch (Exception e) {
				log.error("Connection error: " + e.getMessage());
				errorMessage = e.getMessage();
			}
		}
		if (socketChannel == null) {
			throw new BindException(errorMessage);
		}
	}
	
	private void send(TransferObject to) throws IOException, InterruptedException, ExecutionException, TimeoutException {
		byte[] sendData = TransferUtil.transferObjectToByteArray(to);
		socketChannel.write(ByteBuffer.wrap(sendData)).get(timeout, TimeUnit.MILLISECONDS);
	}
	
	private Object receive(TransferObject to) throws InterruptedException, DataFormatException, SocketExecuteException  {
		Object resultObject = null;

		//if receiveQueue is empty, wait a while, until server response come 
		byte[] receiveData = receiveQueue.poll(timeout, TimeUnit.MILLISECONDS);

		if(receiveData != null){
			resultObject = TransferUtil.byteArrayToReturnObject(receiveData);
		}

		return resultObject;
	}
	
	public void receive(){
		
		//if handleFlag is true, execute
		if(handleFlag){
			byte[] receiveData = bos.toByteArray();
			try {
				if(TransferUtil.isServerCall(receiveData[0])){
					//it is server call
					Worker worker = new Worker(receiveData);
					SocketController.getExecutorService().submit(worker);
				}else{
					//it is server response
					receiveQueue.put(receiveData);
				}
				
			} catch (InterruptedException e) {
				e.printStackTrace();
			}
			handleFlag = false;
			firsttime = true;
			expectReceiveLength = 0;
			receiveLength = 0;
			bos.reset();
		}
		//socketChannel.read(byteBuffer, timeout, TimeUnit.MILLISECONDS, this, aioClientReceiveHandler);
		socketChannel.read(byteBuffer, this, aioClientReceiveHandler);
	}
	
	public void close(){
		try {
			log.debug("Session close to " + socketChannel.getRemoteAddress());
			closed = true;
			//socketChannel.close();
		} catch (Exception e) {
			e.printStackTrace();
		}
	}
	
	public void write(byte[] receiveData){
		try {
			bos.write(receiveData);
		} catch (IOException e) {
			e.printStackTrace();
		}
	}
	
	public void checkOneReceiveDone(int receivedCount, byte[] receiveData){
		receiveLength += receivedCount;
		
		//if finish receive one client request, set handle flag is true
		if(receivedCount < LongClientSession.BUFFER_SIZE){
			handleFlag = true;
		}
		if (receivedCount == LongClientSession.BUFFER_SIZE) {
            if (firsttime) {
                expectReceiveLength = TransferUtil.getExpectReceiveLength(receiveData);
            }
            //if expectReceiveLength equal receiveLength, the server response is done
            if (expectReceiveLength == receiveLength) {
            	handleFlag = true;
            }
        }
		
        firsttime = false;
	}

	private void handleServerCall(byte[] receiveData){
		TransferObject to = new StandardTransferObject();
        try {
	        to = TransferUtil.byteArrayToTransferObject(receiveData);
			execute(to);
		}catch (Exception e) {
			log.error("Callee Class and Method: [" + to.getCalleeClass() + "."+ to.getCalleeMethod() + "]");
			log.error("Handle Receive Data error: " + e);
			close();
		}
	}
	
	private void execute(TransferObject to) throws SocketExecuteException{
		
		try {
			String calleeClassName = to.getCalleeClass();
			String calleeMethod = to.getCalleeMethod();
			
			if(calleeClassName == null){
				throw new SocketExecuteException("No connection with server.");
			}

			//Check default proxy, don't need reflection.
			if(calleeClassName.equals(TransferObject.DEFAULT_PROXY)){
				if(serverCallProxy != null){
					serverCallProxy.execute(to);
				}
			}else{
				Class<?> calleeClass = Class.forName(calleeClassName);
				Object calleeObject = calleeClass.newInstance();

				if (calleeClass == null || calleeObject == null) {
					throw new SocketExecuteException("No connection with server.");
				}

				Method meth = calleeClass.getMethod(calleeMethod, TransferObject.class);
				meth.invoke(calleeObject, to);
			}
			
			

		} catch (Exception e) {
			log.error("Server execute error e: " + e, e);
			throw new SocketExecuteException("Server execute error e: " + e.getMessage());
		} 

	}
	
	class Worker implements Callable<Object>{
		private byte[] receiveData;
		public Worker(byte[] receiveData){
			this.receiveData = receiveData;
		}
		
		@Override
		public Object call() throws Exception {
			handleServerCall(receiveData);
			return 0;
		}
	}
	
	public ByteBuffer getByteBuffer() {
		return byteBuffer;
	}
}

File: history/0.2.0/tatala-client-csharp/src/QiLeYuan/Tatala/socket/client/LongClientSession.cs
using System;
using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Reflection;
using QiLeYuan.Tatala.proxy;
using QiLeYuan.Tatala.socket.to;
using QiLeYuan.Tatala.socket.util;
using QiLeYuan.Tatala.util;
using QiLeYuan.Tools.debug;
using System.Threading;

namespace QiLeYuan.Tatala.socket.client {
    public class LongClientSession {

        private const int QUEUE_SIZE = 10;
        public const string METHOD_CLOSE = "close";

        private String hostIp;
        private int hostPort;
        private int timeout;
        private int retryCount;

        private StateObject mStateObject = new StateObject();
        private bool closed;

        private bool handleFlag = false;
        private bool firsttime = true;
        private int expectReceiveLength = 0;
        private int receiveLength = 0;
        private BlockingQueue receiveQueue = new BlockingQueue(QUEUE_SIZE);
        private DefaultProxy serverCallProxy;

        public delegate void ExecuteDelegate();

        // State object for receiving data from remote device.
        public class StateObject {
            // Client socket.
            public Socket workSocket = null;
            // Size of receive buffer.
            public const int BUFFER_SIZE = 256 * 8;
            // Receive buffer.
            public byte[] buffer = new byte[BUFFER_SIZE];
            // TransferObject
            public TransferObject transferObject = null;
            // Received data
            public MemoryStream memoryStream = new MemoryStream();
        }

        public LongClientSession(String hostIp, int hostPort, int timeout, int retryCount) {
            this.hostIp = hostIp;
            this.hostPort = hostPort;
            this.timeout = timeout;
            this.retryCount = retryCount;
        }

        public Object start(TransferObject to) {
            Object resultObject = null;

            String calleeClass = to.getCalleeClass();
            String calleeMethod = to.getCalleeMethod();

            mStateObject.transferObject = to;

            //check if close method, close client
            if (calleeMethod == METHOD_CLOSE) {
                Close(mStateObject);
                return resultObject;
            }

            //set default server call proxy
            if (to.getServerCallProxy() != null) {
                serverCallProxy = to.getServerCallProxy();
            }

            try {
                if (mStateObject.workSocket == null || !mStateObject.workSocket.Connected || closed) {
                    Connect();
                }
                Send(mStateObject);
                resultObject = Receive(mStateObject);

            } catch (Exception e) {
                Logging.LogError("HostIp and Port: [" + hostIp + ":" + hostPort + "]");
                Logging.LogError("Callee Class and Method: [" + calleeClass + "." + calleeMethod + "]");
                Logging.LogError(e.ToString());
                Close(mStateObject);
                throw e;
            }
            return resultObject;
        }

        private void Connect() {
            //reset close flag is false
            closed = false;

            Socket socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
            IPEndPoint ipEndPoint = new IPEndPoint(IPAddress.Parse(hostIp), hostPort);

            mStateObject.workSocket = socket;

            int retry = retryCount;
            while (true) {
                try {
                    // Connect to the remote endpoint.
                    IAsyncResult result = socket.BeginConnect(ipEndPoint, null, null);
                    bool success = result.AsyncWaitHandle.WaitOne(timeout);
                    if (!success) {
                        retry--;
                        Logging.LogError("Connect timeout");
                        Logging.LogError("Retry count: " + retry);
                    } else {
                        break;
                    }

                } catch (Exception be) {
                    retry--;
                    Logging.LogError(be.Message);
                    Logging.LogError("Retry count: " + retry);
                }
                if (retry <= 0) {
                    Logging.LogError("Connection Error.");
                    throw new Exception("Connection Error.");
                }
            }

            socket.SendTimeout = timeout;
            socket.ReceiveTimeout = timeout;

            //when connect to the server, keep receiving data either server response or server call
            Receive();
        }

        private void Receive() {
            //if handleFlag is true, execute
            if (handleFlag) {
                byte[] receiveData = mStateObject.memoryStream.ToArray();
                if (TransferUtil.isServerCall(receiveData[0])) {
                    //it is server call
                    ExecuteDelegate executeDelegate = delegate() {
                        handleServerCall(receiveData);
                    };
                    executeDelegate.BeginInvoke(null, null);

                    //ThreadPool.QueueUserWorkItem(handleServerCall, receiveData);

                } else {
                    //it is server response
                    receiveQueue.Enqueue(receiveData);
                }

                handleFlag = false;
                firsttime = true;
                expectReceiveLength = 0;
                receiveLength = 0;

                //reset memory stream
                mStateObject.memoryStream.Position = 0;
                mStateObject.memoryStream.SetLength(0);
            }
            Socket socket = mStateObject.workSocket;
            socket.BeginReceive(mStateObject.buffer, 0, StateObject.BUFFER_SIZE, SocketFlags.None, new AsyncCallback(ReceiveCallback), mStateObject);
        }

        private void ReceiveCallback(IAsyncResult ar) {
            // Retrieve the state object and the client socket 
            // from the asynchronous state object.

            StateObject state = (StateObject)ar.AsyncState;

            try {
                Socket socket = state.workSocket;

                if (socket == null || !socket.Connected) {
                    return;
                }

                // Read data from the remote device.
                int bytesReceived = socket.EndReceive(ar);

                if (bytesReceived > 0) {
                    // There might be more data, so store the data received so far.
                    state.memoryStream.Write(mStateObject.buffer, 0, bytesReceived);

                    //check whether one receive from server is done
                    CheckOneReceiveDone(bytesReceived, mStateObject.buffer);

                    Receive();
                } else {
                    //socket may colsed, due to network issue
                    Close(state);
                }
            } catch (Exception e) {
                Logging.LogError(e.ToString());
                Close(state);
            }
        }

        private void Send(StateObject stateObject) {
            TransferObject to = stateObject.transferObject;
            Socket socket = stateObject.workSocket;

            byte[] sendData = TransferUtil.transferObjectToByteArray(to);

            //socket.Send(sendData);

            // Begin sending the data to the remote device.     
            socket.BeginSend(sendData, 0, sendData.Length, SocketFlags.None, new AsyncCallback(SendCallback), socket);
        }

        private void SendCallback(IAsyncResult ar) {
            try {
                // Retrieve the socket from the state object.     
                Socket socket = (Socket)ar.AsyncState;
                // Complete sending the data to the remote device.     
                socket.EndSend(ar);
            } catch (Exception e) {
                Logging.LogError("Send error: "+e.ToString());
            }
        }

        private Object Receive(StateObject stateObject) {
            Object resultObject = null;

            //if return void, don't call socket receive
            if (stateObject.transferObject.getReturnType() == TransferObject.DATATYPE_VOID) {
                return resultObject;
            }

            //if receiveQueue is empty, wait a while, until server response come 
            byte[] receiveData = (byte[])receiveQueue.Dequeue(timeout);

            if (receiveData != null) {
                resultObject = TransferUtil.byteArrayToReturnObject(receiveData);
            } else {
                Logging.LogError("receive timeout return null");
            }
            return resultObject;
        }

        public void CheckOneReceiveDone(int receivedCount, byte[] receiveData) {
            receiveLength += receivedCount;

            //if finish receive one client request, set handle flag is true
            if (receivedCount < StateObject.BUFFER_SIZE) {
                handleFlag = true;
            }
            if (receivedCount == StateObject.BUFFER_SIZE) {
                if (firsttime) {
                    expectReceiveLength = TransferUtil.getExpectReceiveLength(receiveData);
                }
                //if expectReceiveLength equal receiveLength, the server response is done
                if (expectReceiveLength == receiveLength) {
                    handleFlag = true;
                }
            }

            firsttime = false;
        }

        private void Close(StateObject stateObject) {
            closed = true;
            if (stateObject.workSocket.Connected) {
                stateObject.workSocket.Shutdown(SocketShutdown.Both);
            }
            stateObject.workSocket.Close();
        }

        private void handleServerCall(object receiveObj) {
            byte[] receiveData = (byte[])receiveObj;
            TransferObject to = new StandardTransferObject();
            try {
                to = TransferUtil.byteArrayToTransferObject(receiveData);
                execute(to);
            } catch (Exception e) {
                Logging.LogError("Callee Class and Method: [" + to.getCalleeClass() + "." + to.getCalleeMethod() + "]");
                Logging.LogError("Handle Receive Data error: " + e);
                Close(mStateObject);
            }
        }

        private void execute(TransferObject to) {

            String calleeClassName = to.getCalleeClass();

            //Check default proxy, don't need reflection.
            if (calleeClassName.Equals(TransferObject.DEFAULT_PROXY)) {
                if (serverCallProxy != null) {
                    serverCallProxy.execute(to);
                }
            } else {
                String calleeMethod = to.getCalleeMethod();
                Type type = Type.GetType(calleeClassName);
                Object instance = Activator.CreateInstance(type);
                MethodInfo method = type.GetMethod(calleeMethod);
                method.Invoke(instance, new object[] { to });
            }

        }

    }
}
